### ISSUE TITLE
Retrieve the LoaderExceptions property for more information

### ISSUE THREAD

**Issue by MisinformedDNA** [Azure/azure-functions-host#2864]
> [5/17/2018 6:54:51 PM] Reading host configuration file 'C:\Users\XXXX\Documents\Visual Studio 2017\Projects\SomeProject\SomeApp.Functions\bin\Debug\netstandard2.0\host.json'
> [5/17/2018 6:54:51 PM] Host configuration file read:
> [5/17/2018 6:54:51 PM] {}
> [5/17/2018 6:54:51 PM] Starting Host (HostId=pc0gpvx6-1892505880, InstanceId=8728c5e1-c5f6-4681-a6c8-e7114fb75ac9, Version=2.0.11651.0, ProcessId=24804, AppDomainId=1, Debug=False, ConsecutiveErrors=0, StartupCount=1, FunctionsExtensionVersion=)
> [5/17/2018 6:54:51 PM] Unable to configure java worker. Could not find JAVA_HOME app setting.
> [5/17/2018 6:54:51 PM]
> [5/17/2018 6:54:51 PM] Could not configure language worker Java.
> [5/17/2018 6:54:51 PM]
> Warning: Only got partial types from assembly: SomeApp.Functions, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
> Exception message: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
>    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.Assembly.GetTypes()
>    at Microsoft.Azure.WebJobs.Host.Indexers.DefaultTypeLocator.FindTypes(Assembly assembly, IEnumerable`1 extensionAssemblies) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Indexers\DefaultTypeLocator.cs:line 120
> [5/17/2018 6:54:53 PM] Generating 0 job function(s)
> [5/17/2018 6:54:53 PM] Found the following functions:
> [5/17/2018 6:54:53 PM] SomeApp.Functions.SomeFunction.RunAsync
> [5/17/2018 6:54:53 PM] SomeApp.Functions.SomeOtherFunction.Run
> [5/17/2018 6:54:53 PM]
> [5/17/2018 6:54:53 PM] Host initialized (2377ms)
> Listening on http://localhost:7071/
> Hit CTRL-C to exit...
> [5/17/2018 6:54:54 PM] Host started (2925ms)
> [5/17/2018 6:54:54 PM] Job host started
> [5/17/2018 6:54:54 PM] The following 2 functions are in error:
> [5/17/2018 6:54:54 PM] SomeFunction: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
> [5/17/2018 6:54:54 PM] SomeOtherFunction: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
> [5/17/2018 6:54:54 PM]
> [5/17/2018 6:54:54 PM]
> [5/17/2018 6:54:54 PM] Host lock lease acquired by instance ID '00000000000000000000000018C0B5EE'.

This is happening on my local machine.

#### Expected behavior

No errors

#### Actual behavior

Provide a description of the actual behavior observed.

> Exception message: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
>    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.Assembly.GetTypes()
>    at Microsoft.Azure.WebJobs.Host.Indexers.DefaultTypeLocator.FindTypes(Assembly assembly, IEnumerable`1 extensionAssemblies) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Indexers\DefaultTypeLocator.cs:line 120

#### Related information 

<details>
<summary>SomeOtherFunction.cs</summary>

```csharp
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.Azure.WebJobs.Host;

namespace SomeApp
{
    public static class SomeOtherFunction
    {
        [FunctionName(nameof(SomeOtherFunction))]
        public static IActionResult Run([HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)]HttpRequest req, TraceWriter log)
        {
            log.Info("C# HTTP trigger function processed a request.");
            return null;
        }
    }
}
```
</details>
<details>
<summary>SomeApp.csproj</summary>

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <AzureFunctionsVersion>v2</AzureFunctionsVersion>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
  
  <ItemGroup>
    <PackageReference Include="HtmlAgilityPack" Version="1.8.2" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.13" />
    <PackageReference Include="morelinq" Version="2.10.0" />
  </ItemGroup>
  
  <ItemGroup>
    <ProjectReference Include="..\SomeDependency\SomeDependency.csproj" />
  </ItemGroup>

  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
    <None Update="Wookieepedia.html">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
  </ItemGroup>
  
</Project>
```
</details>

* C#
* HttpTrigger
* Microsoft.NET.Sdk.Functions 1.0.13
* VS 2017 15.7
* Azure Functions and Web JobsTools 15.0.40502.0

**Comment by MisinformedDNA** [comment]
I've narrowed it down to a code file. If this file exists, I get the error above. If I remove the file, everything works.

```csharp
using SomeDependency;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Table;
using MoreLinq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
```

Also, even though the functions are never run, could it have something to do with me trying to access a file?

```csharp
        public static string Load()
        {
            return File.ReadAllText("ParseThis.html");
        }
```

**Comment by MisinformedDNA** [comment]
Nope, I removed that function and the error persists.

**Comment by MisinformedDNA** [comment]
OK, I finally made some progress by debugging into the Function code. Here is the LoaderException:

> Could not load file or assembly 'Microsoft.WindowsAzure.Storage, Version=9.1.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)

**Comment by nathankitchen** [comment]
This is driving me crazy too. I’ve burned a good 8 hours into trying to work out which assembly I have out of sync. Using Functions beta/netstandard2 so I expect a rough ride, kind of silly at the moment though.

Problem is my functions run flawlessly locally but fail with this same message after publishing. I’ve established I’m running the latest version locally but the Azure version is newer.

All I want to do is get the LoaderExceptions, I can probably figure the rest out myself. Tried intercepting App Domain events but if they’re working I can’t find a way to write them out to logs.

Please can the runtime be updated to specifically handle ReflectionTypeLoadExceptions at the point that the runtime prints “The following n functions are in error:” and enumerate the loader exceptions to the log?

**Comment by fabiocav** [comment]
@MisinformedDNA that version shouldn't be what is expected in 2.0, as  it currently uses 8.x

This is an issue that is being mitigated in waves in 2.0, with the first wave of assembly loading and binding enhancements having landed in the last release. We're waiting on core SDK work to be completed in order to enable the same behavior for the storage SDK assemblies (currently, we're explicitly locking that to the runtime version in order to prevent runtime failures). That work should be completed soon.

@nathankitchen the Core Tools update is in progress. You can run the latest bits locally by explicitly installing using npm (`npm i -g azure-functions-core-tools@core`), and the VS tools should be picking that up soon.

**Comment by MisinformedDNA** [comment]
Here's what I see in my output:

> Hosting environment: Production
> Content root path: C:\Users\XXXX\Documents\Visual Studio 2017\Projects\XXXX\XXXX.Functions\bin\Debug\netstandard2.0
> Now listening on: http://localhost:7071
> Application started. Press Ctrl+C to shut down.
> [5/22/2018 7:05:22 PM] Reading host configuration file 'C:\Users\XXXX\Documents\Visual Studio 2017\Projects\XXXX\XXXX.Functions\bin\Debug\netstandard2.0\host.json'
> [5/22/2018 7:05:22 PM] Host configuration file read:
> [5/22/2018 7:05:22 PM] {}
> [5/22/2018 7:05:22 PM] Starting Host (HostId=pc0gpvx6-1892505880, InstanceId=8533e06a-aec3-42fb-b2a8-b1a4e171b64b, Version=2.0.11776.0, ProcessId=4316, AppDomainId=1, Debug=False, ConsecutiveErrors=0, StartupCount=1, FunctionsExtensionVersion=)
> [5/22/2018 7:05:22 PM] FUNCTIONS_WORKER_RUNTIME is specified, only dotnet will be enabled
> Warning: Only got partial types from assembly: CanonChecklist.Functions, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
> Exception message: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
>    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.Assembly.GetTypes()
>    at Microsoft.Azure.WebJobs.Host.Indexers.DefaultTypeLocator.FindTypes(Assembly assembly, IEnumerable`1 extensionAssemblies) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Indexers\DefaultTypeLocator.cs:line 120
> [5/22/2018 7:05:47 PM] Generating 0 job function(s)

But using SourceLink, I see the 9.x error:

![image](https://user-images.githubusercontent.com/1784452/40385274-a819c740-5dcb-11e8-873a-79e7f260c863.png)

If I change to 8.x latest (i.e. 8.7.0), I get the same error. If I change to 8.2.1, it works.

This is a difficult bug to debug. Hopefully, things will be better soon with the bindng enhancements in place.

**Comment by fabiocav** [comment]
This has been addressed by the fix put in place for #1722

**Comment by chmarti** [comment]
@fabiocav how exactly was this addressed? I'm on the latest version of dotnet, visual studio, functions host / tools (azure-functions-core-tools  2.0.1-beta.28) and I'm still stuck in a situation where these TypeLoad errors are really difficult to debug. What I guess I'm not understanding is:
1. Why is a certain version of Azure storage required? How do I know which one it is?
2. Why is the actual type failing to be loaded not displayed.
3. Given I'm having this situation on the latest version, why is this bug closed? If #1722 fixed this and it was closed in September, why are we still in this state?

**Comment by chmarti** [comment]
@fabiocav thanks for the reply. 

npm has 2.0.1-beta.29 now and I am currently running that version.

My function is using an Azure Queue trigger and an Azure Blob input binding. This was all working locally and in Azure until about a week ago. I assume there was an Azure deployment that changed the version of Azure Storage dependency and that is why I started getting the LoaderExceptions.

When I run "func host start" from the command line, I see the version: Version=2.0.11857.0. I am not seeing any LoaderExceptions exceptions when I run this way, but I'm struggling to get the function running properly without help of the VS Extension.

I verified that I have the latest version of VS Community (15.7.3) and the Azure Functions and Web Jobs Tools Extension (15.0.40502.0). When I try to debug by function app from within VS the version running is: Version=2.0.11776.0. When running this way, I obviously still see the LoaderExceptions.

Is there a way to get the VS Extension to use the latest version of azure-functions-core-tools@core installed via npm? For the record I installed it with -g and have restarted VS since, but it is still picking up the old version somewhere.
